### PR TITLE
ROC-2069 Defendant change defence response option

### PR DESCRIPTION
**Pull request by rattur**
Defendant response page options and flow change
 
### JIRA link ###

https://tools.hmcts.net/jira/browse/ROC-2069

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

**Comment by timja-kainos** [comment]
conflicts

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+{% set heading = t("I reject all of the claim") %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+
+      <form method="post" novalidate>
+        {{ csrfProtection(csrf) }}
+
+        {% set error = form.errorFor('type') %}
```
Comment (timja-kainos) [inline src/main/features/response/views/reject-all-of-claim.njk]: why is this here

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        <div class="form-group {% if error %} form-group-error {% endif %}">
+          <fieldset id="type[label]">
```
Comment (timja-kainos) [inline src/main/features/response/views/reject-all-of-claim.njk]: pretty sure radiogroup comes with a fieldset

File: src/main/features/response/views/reject-part-of-claim.njk [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+{% set heading = t("I reject part of the claim") %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+
+      <form method="post" novalidate>
+        {{ csrfProtection(csrf) }}
+        {% set error = form.errorFor('type') %}
```
Comment (timja-kainos) [inline src/main/features/response/views/reject-part-of-claim.njk]: this shouldn't be here

File: src/main/features/response/views/reject-part-of-claim.njk [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        <div class="form-group {% if error %} form-group-error {% endif %}">
+          <fieldset id="type[label]">
+            <legend class="visually-hidden">{{ t("I reject part of the claim") }}</legend>
+            {% if error %}<span class="error-message">{{ t(error) }}</span>{% endif %}
```
Comment (timja-kainos) [inline src/main/features/response/views/reject-part-of-claim.njk]: down to here from above comment shouldn't be here

File: src/main/features/response/routes/response-type.ts [hunk 5]
```diff
@@ -36,9 +36,9 @@ export default express.Router()
         await ResponseDraftMiddleware.save(res, next)
 
         if (ResponseType.OWE_NONE === form.model.type) {
-          res.redirect(Paths.defenceOptionsPage.evaluateUri({ externalId: externalId }))
```
Comment (timja-kainos) [inline src/main/features/response/routes/response-type.ts]: this check isn't right, there's 3 pages it could go to:
part admission
full rejection
task-list

I just pulled your branch and the full admission button is going to partial admission

File: src/main/features/response/routes/check-and-send.ts [hunk 6]
```diff
@@ -103,19 +102,16 @@ export default express.Router()
         switch (responseType) {
           case ResponseType.OWE_NONE:
             if (defendantIsCounterClaiming(user)) {
-              res.redirect(Paths.counterClaimPage.evaluateUri({ externalId: user.claim.externalId }))
+              res.redirect(Paths.defenceRejectAllOfClaimPage.evaluateUri({ externalId: user.claim.externalId }))
```
Comment (timja-kainos) [inline src/main/features/response/routes/check-and-send.ts]: this isn't right its going to a page with radio buttons, it should still go to the guidance page until we can handle it.

File: src/main/features/response/routes/check-and-send.ts [hunk 7]
```diff
@@ -103,19 +102,16 @@ export default express.Router()
...
               return
             }
             break
           case ResponseType.OWE_SOME_PAID_NONE:
-          case ResponseType.OWE_ALL_PAID_SOME:
-            res.redirect(Paths.partialAdmissionPage.evaluateUri({ externalId: user.claim.externalId }))
+            res.redirect(Paths.defenceRejectPartOfClaimPage.evaluateUri({ externalId: user.claim.externalId }))
```
Comment (timja-kainos) [inline src/main/features/response/routes/check-and-send.ts]: this isn't right its going to a page with radio buttons, it should still go to the guidance page until we can handle it.

File: src/main/features/response/routes/check-and-send.ts [hunk 8]
```diff
@@ -103,19 +102,16 @@ export default express.Router()
...
             return
           case ResponseType.OWE_ALL_PAID_NONE:
-            res.redirect(Paths.fullAdmissionPage.evaluateUri({ externalId: user.claim.externalId }))
+            res.redirect(Paths.defenceRejectPartOfClaimPage.evaluateUri({ externalId: user.claim.externalId }))
```
Comment (timja-kainos) [inline src/main/features/response/routes/check-and-send.ts]: this isn't right its going to a page with radio buttons, it should still go to the guidance page until we can handle it.

File: src/main/features/response/views/response-type.njk [hunk 9]
```diff
@@ -16,30 +16,23 @@
           <legend class="visually-hidden">{{ t("Do you owe the money claimed?") }}</legend>
           {% if error %}<span class="error-message">{{ t(error) }}</span>{% endif %}
 
-          {{ radioOption(form = form, name = 'type[value]', value = 'OWE_ALL_PAID_ALL',
-            label = t('I’ve already paid'), bold = true,
-            hint = t("You need to submit a defence explaining how much you’ve paid the claimant and when you paid.")
-          )}}
+          <p>{{ t("You can make a counterclaim against the claimant after you submit your response.") }}</p>
```
Comment (timja-kainos) [inline src/main/features/response/views/response-type.njk]: you are missing the needed configuration for the analytics sender to send the data to google analytics
see this PR:
https://github.com/hmcts/cmc-citizen-frontend/pull/58/files

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        },
+        {
+          label: "I dispute the claim",
+          value: "dispute",
+          hint: "",
+          bold: true
+        },
+        {
+          label: 'Defend now and also make a counterclaim',
+          value: "counterClaim",
```
Comment (geek-soft) [inline src/main/features/response/views/reject-all-of-claim.njk]: I do not see this being used anywhere in the application. How does is fit into the counter claim feature we already have? Either start using it or delete.

File: src/main/features/response/routes/reject-part-of-claim.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import User from 'idam/user'
+
+function renderView (form: Form<RejectPartOfClaim>, res: express.Response) {
+  res.render(Paths.defenceRejectPartOfClaimPage.associatedView, {
+    form: form
+  })
+}
+
+export default express.Router()
+  .get(Paths.defenceRejectPartOfClaimPage.uri, ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
```
Comment (geek-soft) [inline src/main/features/response/routes/reject-part-of-claim.ts]: That route has to be protected by guard so it is impossible to access it unless somebody selected `OWE_SOME_PAID_NONE` on the response type page.

Remember to cover both GET and POST handlers and reflect that fact in tests as well.

File: src/main/features/response/routes/reject-all-of-claim.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+import User from 'idam/user'
+
+function renderView (form: Form<RejectAllOfClaim>, res: express.Response) {
+  res.render(Paths.defenceRejectAllOfClaimPage.associatedView, {
+    form: form
+  })
+}
+
+export default express.Router()
+  .get(Paths.defenceRejectAllOfClaimPage.uri, ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
```
Comment (geek-soft) [inline src/main/features/response/routes/reject-all-of-claim.ts]: That route has to be protected by guard so it is impossible to access it unless somebody selected `OWE_NONE` on the response type page.

Remember to protected both GET and POST handlers and reflect that fact in tests as well.

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        {% set error = form.errorFor('type') %}
+        <div class="form-group {% if error %} form-group-error {% endif %}">
+          <fieldset id="type[label]">
+            <legend class="visually-hidden">{{ t("I reject all of the claim") }}</legend>
+            {% if error %}<span class="error-message">{{ t(error) }}</span>{% endif %}
+        {{
+        radioGroup(
+        label = "Why do you reject all the claim?",
+        name = "option",
+        options = [
```
[the remark this answers is not in the document]
Comment (rattur) [reply src/main/features/response/views/reject-all-of-claim.njk]: Yes I will address this, need to change statement of truth template too

**Comment by geek-soft** [comment]
PR needs merge with master. This RP is build on out-of-sync master which does not even have changed structure of the drafts.

File: src/main/features/response/routes/response-type.ts [hunk 14]
```diff
@@ -36,9 +35,13 @@ export default express.Router()
         await DraftService.save(user.responseDraft, user.bearerToken)
 
         if (ResponseType.OWE_NONE === form.model.type) {
```
Comment (geek-soft) [inline src/main/features/response/routes/response-type.ts]: Using `switch` here makes more sense, it should also make code a bit nicer / easier to read.

File: src/main/features/response/draft/responseDraft.ts [hunk 15]
```diff
@@ -32,6 +36,8 @@ export class ResponseDraft extends DraftDocument implements Serializable<Respons
       if (input.qualifiedStatementOfTruth) {
         this.qualifiedStatementOfTruth = new QualifiedStatementOfTruth().deserialize(input.qualifiedStatementOfTruth)
       }
+      this.rejectPartOfClaim = new RejectPartOfClaim(input.rejectPartOfClaim && input.rejectPartOfClaim.option)
+      this.rejectAllOfClaim = new RejectPartOfClaim(input.rejectAllOfClaim && input.rejectAllOfClaim.option)
```
Comment (geek-soft) [inline src/main/features/response/draft/responseDraft.ts]: Type does not match here, you should have used `RejectAllOfClaim` here.

File: src/main/features/response/guards/oweSomePaidNoneResponseRequiredGuard.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import * as express from 'express'
+
+import { ResponseType } from 'response/form/models/responseType'
+import { Paths } from 'response/paths'
+import User from 'idam/user'
+
+export default class OweSomePaidNoneResponseRequiredGuard {
```
Comment (geek-soft) [inline src/main/features/response/guards/oweSomePaidNoneResponseRequiredGuard.ts]: Please do not use default import. Using default imports make refactoring / code maintenance slightly more difficult. Please change to named import.

Comment (geek-soft) [inline src/main/features/response/guards/oweSomePaidNoneResponseRequiredGuard.ts]: That class seems not to be used at all. What is the point of having guard / a class which is not used?

File: src/main/features/response/tasks/oweMoneyTask.ts [hunk 17]
```diff
@@ -7,6 +7,13 @@ export class OweMoneyTask {
       return false
     }
 
+    if (responseDraft.response.type !== ResponseType.OWE_ALL_PAID_NONE) {
```
Comment (geek-soft) [inline src/main/features/response/tasks/oweMoneyTask.ts]: That does not even change anything. I can still choose an option on response type page, be taken to the response details screen, jump to task list without doing anything and see COMPLETED. That is not CORRECT. Task is completed only when all necessary / mandatory details has been captured.

Please correct the logic and reflect it in the unit test.

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+{% set heading = t("I reject all of the claim") %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+
+      <form method="post" novalidate>
+        {{ csrfProtection(csrf) }}
+
+        <div class="form-group {% if error %} form-group-error {% endif %}">
```
Comment (geek-soft) [inline src/main/features/response/views/reject-all-of-claim.njk]: That is not needed as `radioGroup` macro does it for us. Please check what `radioGroup` is doing and remove unnecessary markup.

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+
+            <legend class="visually-hidden">{{ t("I reject all of the claim") }}</legend>
+            {% if error %}<span class="error-message">{{ t(error) }}</span>{% endif %}
```
Comment (geek-soft) [inline src/main/features/response/views/reject-all-of-claim.njk]: This is not needed as well.

File: src/main/features/response/views/reject-all-of-claim.njk (hunk 13 above)
[the remark this answers is not in the document]
Comment (geek-soft) [reply src/main/features/response/views/reject-all-of-claim.njk]: You did say that you will address it however it is still not addressed.

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+        {
+          label: 'Defend now and also make a counterclaim',
+          value: "counterClaim",
+          hint: 'You’ll need to complete the defence and counterclaim form.'
+        }
+        ],
+        form = form
+        )
+        }}
+          <div class="form-group">
```
Comment (geek-soft) [inline src/main/features/response/views/reject-all-of-claim.njk]: We do not put buttons into `form-group` div. Any reason you did it? If not please remove.

File: src/main/features/response/views/reject-part-of-claim.njk [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+{% set heading = t("I reject part of the claim") %}
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+
+      <form method="post" novalidate>
+        {{ csrfProtection(csrf) }}
+        <div class="form-group {% if error %} form-group-error {% endif %}">
```
Comment (geek-soft) [inline src/main/features/response/views/reject-part-of-claim.njk]: The same here and other places in the file.

File: src/main/views/includes/scripts.njk [hunk 22]
```diff
@@ -6,7 +6,7 @@
 <script src="{{ asset_paths['js'] }}/application.js"></script>
 <script src="{{ asset_paths['js'] }}/detectJavascript.js"></script>
 <script src="{{ asset_paths['js'] }}/analytics.js"></script>
-<script src="{{ asset_paths['js'] }}/analytics-custom-event-sender.js"></script>
+<script src="{{ asset_paths['js'] }}/analytics-custom-event-sender.js?2017-10-11"></script>
```
Comment (geek-soft) [inline src/main/views/includes/scripts.njk]: Should not be needed but preferred to add it in case. In the future asset fingerprinting should take care of it.

File: src/test/a11y/mocks.ts [hunk 23]
```diff
@@ -63,3 +63,11 @@ mock('ccj/guards/ccjGuard', {
 mock('ccj/guards/individualDateOfBirthGuard', {
   'IndividualDateOfBirthGuard': justForwardRequestHandler
 })
+
+mock('response/guards/guardFactory', {
+  GuardFactory: {
```
Comment (geek-soft) [inline src/test/a11y/mocks.ts]: In the future this generic guard factory should help us keeping route specific guards close to the routes and extract guards to separate files only when these need to be shared across many routes

File: src/main/features/response/form/models/rejectAllOfClaim.ts [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+      RejectAllOfClaimOption.DISPUTE,
+      RejectAllOfClaimOption.COUNTER_CLAIM
+    ]
+  }
+
+  static except (option: string): string[] {
+    if (option === undefined) {
+      throw new Error('Option is required')
+    }
+    return RejectAllOfClaimOption.all().filter(_ => _ !== option)
```
Comment (rlewan) [inline src/main/features/response/form/models/rejectAllOfClaim.ts]: Can we use a word as an argument name instead of `_`?

File: src/main/features/response/routes/reject-all-of-claim.ts [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+export default express.Router()
+  .get(
+    Paths.defenceRejectAllOfClaimPage.uri,
+    GuardFactory.create(isRequestAllowed, accessDeniedCallback),
+    ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+      renderView(new Form(res.locals.user.responseDraft.document.rejectAllOfClaim), res)
+    }))
+  .post(
+    Paths.defenceRejectAllOfClaimPage.uri,
+    GuardFactory.create(isRequestAllowed, accessDeniedCallback),
```
Comment (rlewan) [inline src/main/features/response/routes/reject-all-of-claim.ts]: This guard can be created once as a constant and then passed in as a parameter.

File: src/main/features/response/routes/reject-part-of-claim.ts [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+export default express.Router()
+  .get(
+    Paths.defenceRejectPartOfClaimPage.uri,
+    GuardFactory.create(isRequestAllowed, accessDeniedCallback),
+    ErrorHandling.apply(async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+      renderView(new Form(res.locals.user.responseDraft.document.rejectPartOfClaim), res)
+    }))
+  .post(
+    Paths.defenceRejectPartOfClaimPage.uri,
+    GuardFactory.create(isRequestAllowed, accessDeniedCallback),
```
Comment (rlewan) [inline src/main/features/response/routes/reject-part-of-claim.ts]: Same here, the guard can be created once.

File: src/main/features/response/routes/response-type.ts [hunk 27]
```diff
@@ -35,10 +35,20 @@ export default express.Router()
...
+            res.redirect(Paths.defenceRejectAllOfClaimPage.evaluateUri({ externalId: externalId }))
+            break
+          case ResponseType.OWE_SOME_PAID_NONE:
+            res.redirect(Paths.defenceRejectPartOfClaimPage.evaluateUri({ externalId: user.claim.externalId }))
+            break
+          case ResponseType.OWE_ALL_PAID_NONE:
+            res.redirect(Paths.taskListPage.evaluateUri({ externalId: user.claim.externalId }))
+            break
+          default:
+            next(new Error('Unknown response type: ' + responseType))
```
Comment (rlewan) [inline src/main/features/response/routes/response-type.ts]: Backtics can be used here to interpolate the string.

File: src/main/features/response/routes/reject-part-of-claim.ts [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+
+import { ResponseType } from 'response/form/models/responseType'
+import { RejectPartOfClaim } from 'response/form/models/rejectPartOfClaim'
+import { DraftService } from 'common/draft/draftService'
+import { ErrorHandling } from 'common/errorHandling'
+import User from 'idam/user'
+import { GuardFactory } from 'response/guards/guardFactory'
+
+function isRequestAllowed (res: express.Response): boolean {
+  return res.locals.user.responseDraft.document.response !== undefined
```
Comment (yarpo) [inline src/main/features/response/routes/reject-part-of-claim.ts]: why don't we assign:

```
const document: responseDraft = res.locals.user.responseDraft.document
```

?

File: src/main/features/response/views/reject-all-of-claim.njk [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+        <legend class="visually-hidden">{{ t('I reject all of the claim') }}</legend>
+
+        {{
+          radioGroup(
+          label = 'Why do you reject all the claim?',
+          name = 'option',
+          options = [
+            {
+              label: 'I’ve already paid the claim in full',
+              value: 'alreadyPaid'
```
Comment (rlewan) [inline src/main/features/response/views/reject-all-of-claim.njk]: Seeing this (not only here but on all such pages as well) I have a feeling we should export all constants classes as nunjucks locals and use them instead of using magic values.

File: src/main/features/response/views/check-and-send.njk [hunk 30]
```diff
@@ -39,18 +39,16 @@
...
-      {% endif %}
-
-      {{ submitButton('Submit response') }}
-      <input type="hidden" name="type" value="{{ form.model.type }}">
-    </div>
+    {{
+      statementOfTruth(
+        form,
+        label = 'I believe that the facts stated in this claim are true',
+        isQualified = (signatureType == 'qualified')
```
Comment (yarpo) [inline src/main/features/response/views/check-and-send.njk]: `===` ?

File: src/main/public/js/analytics-custom-event-sender.js [hunk 31]
```diff
@@ -13,6 +13,15 @@ $(function () {
     ga('send', 'event', eventCategory, eventAction, eventLabel)
   }
 
+  function escape (value) {
+    if (!value) {
```
Comment (rlewan) [inline src/main/public/js/analytics-custom-event-sender.js]: We should also check that the value is a string.

File: src/test/features/response/draft/responseDraft.ts [hunk 32]
```diff
@@ -47,33 +47,53 @@ describe('ResponseDraft', () => {
...
+  })
 
-        expect(responseDraftModel.requireDefence()).to.be.eq(false)
-      })
-      it('should return false when response type is OWE_NONE and counter claim is true', () => {
-        const responseDraftModel: ResponseDraft = new ResponseDraft()
-        responseDraftModel.response = new Response(ResponseType.OWE_NONE)
-        responseDraftModel.counterClaim = new CounterClaim(true)
+  describe('requireDefence', () => {
+    it('should return false when no response type set', () => {
```
Comment (rlewan) [inline src/test/features/response/draft/responseDraft.ts]: Either this method is named poorly or there is a bug in its logic. I would expect a method named `requireDefence` to return true if the response is not set.

File: src/main/features/response/views/reject-all-of-claim.njk (hunk 29 above)
Comment (geek-soft) [reply src/main/features/response/views/reject-all-of-claim.njk]: Agree, asked @rattur for the same but it got lower priority.

File: src/main/features/response/form/models/rejectAllOfClaim.ts (hunk 24 above)
Comment (geek-soft) [reply src/main/features/response/form/models/rejectAllOfClaim.ts]: Done

File: src/main/features/response/routes/reject-all-of-claim.ts (hunk 25 above)
Comment (geek-soft) [reply src/main/features/response/routes/reject-all-of-claim.ts]: Done

File: src/main/features/response/routes/reject-part-of-claim.ts (hunk 26 above)
Comment (geek-soft) [reply src/main/features/response/routes/reject-part-of-claim.ts]: Done

File: src/main/features/response/routes/response-type.ts (hunk 27 above)
Comment (geek-soft) [reply src/main/features/response/routes/response-type.ts]: Done

File: src/main/features/response/views/check-and-send.njk (hunk 30 above)
Comment (geek-soft) [reply src/main/features/response/views/check-and-send.njk]: Done

File: src/main/features/response/routes/reject-part-of-claim.ts (hunk 28 above)
Comment (geek-soft) [reply src/main/features/response/routes/reject-part-of-claim.ts]: We could, but I would like to create type definition more so that `res.locals.user.responseDraft.document` gives us right type. Having this we would not need to reassign to variable so often.

File: src/main/public/js/analytics-custom-event-sender.js (hunk 31 above)
Comment (geek-soft) [reply src/main/public/js/analytics-custom-event-sender.js]: Done.
